### PR TITLE
binlog encode implementation

### PR DESCRIPTION
**Pull request by chaychen2005**

### CODE REVIEW

File: libstorage/BinLogHandler.cpp [hunk 1]
```diff
@@ -99,4 +101,127 @@ bool BinLogHandler::getMissingBlocksFromBinLog(
...
+        writeUINT64(buffer, entry->getID());
+        uint8_t status = entry->getStatus();
+        BINLOG_HANDLER_LOG(TRACE) << "entry id:" << entry->getID()
+                                  << ", status:" << (uint32_t)status;
+        buffer.insert(buffer.end(), (byte*)&status, (byte*)&status + 1);
+        for (size_t k = 0; k < vecField.size(); k++)
+        {
+            BINLOG_HANDLER_LOG(TRACE) << "key:" << vecField[k];
+            auto fieldIt = entry->find(vecField[k]);
+            if (fieldIt != entry->end())
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp]: please handle key doesn't exist.

File: libstorage/BinLogHandler.cpp [hunk 2]
```diff
@@ -99,4 +101,127 @@ bool BinLogHandler::getMissingBlocksFromBinLog(
...
+}
+
+void BinLogHandler::writeString(bytes& buffer, const std::string& str)
+{
+    uint32_t strLen = htonl(str.length());
+    buffer.insert(buffer.end(), (byte*)&strLen, (byte*)&strLen + 4);
+    buffer.insert(buffer.end(), str.begin(), str.end());
+}
+
+void BinLogHandler::readUINT32(const bytes& buffer, uint32_t& ui, uint32_t& offset)
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp]:
```suggestion

```
use `ntohl` is enough

File: libstorage/BinLogHandler.cpp [hunk 3]
```diff
@@ -99,4 +101,127 @@ bool BinLogHandler::getMissingBlocksFromBinLog(
...
+{
+    ui = ntohl(*((uint32_t*)&buffer[offset]));
+    offset += 4;
+}
+
+void BinLogHandler::readUINT64(const bytes& buffer, uint64_t& ui, uint32_t& offset)
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp]:
```suggestion
uint64_t BinLogHandler::ntoll(uint64_t& ui)
```

File: libstorage/BinLogHandler.cpp (hunk 1 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Done.

File: libstorage/BinLogHandler.cpp [hunk 4]
```diff
@@ -99,4 +101,123 @@ bool BinLogHandler::getMissingBlocksFromBinLog(
...
+void BinLogHandler::writeUINT32(bytes& buffer, uint32_t ui)
+{
+    uint32_t ui_tmp = htonl(ui);
+    buffer.insert(buffer.end(), (byte*)&ui_tmp, (byte*)&ui_tmp + 4);
+}
+
+void BinLogHandler::writeUINT64(bytes& buffer, uint64_t ui)
+{
+    writeUINT32(buffer, ((ui >> 32) & 0xffffffff));
+    writeUINT32(buffer, (ui & 0xffffffff));
```
Comment (morebtcg) [inline libstorage/BinLogHandler.cpp]: wrong impl

File: libstorage/BinLogHandler.cpp [hunk 5]
```diff
@@ -99,4 +101,123 @@ bool BinLogHandler::getMissingBlocksFromBinLog(
...
+    uint32_t ui = ntohl(*((uint32_t*)&buffer[offset]));
+    offset += 4;
+    return ui;
+}
+
+uint64_t BinLogHandler::readUINT64(const bytes& buffer, uint32_t& offset)
+{
+    uint32_t uih = readUINT32(buffer, offset);
+    uint32_t uil = readUINT32(buffer, offset);
+    uint64_t ui = (uint64_t)uih << 32 | uil;
```
Comment (morebtcg) [inline libstorage/BinLogHandler.cpp]: wrong impl

Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Now it is processed separately according to the situation of the machine.

File: libstorage/BinLogHandler.cpp (hunk 4 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Now it is processed separately according to the situation of the machine.

**Pull request merged**